### ISSUE TITLE
VulkanReplayConsumerBase::OverrideGetEventStatus can get easily on a infinite loop

### ISSUE THREAD

**Issue by infapi00** [LunarG/gfxreconstruct#518]
Right now the code of that method includes this loop:
```

   do
    {
        result = func(device, event);
    } while ((original_result == VK_EVENT_SET) && (result == VK_EVENT_RESET));
```

That as far as I understand, the idea is asking for the event status until it gets the same value that it was recorded.

But if that never happens, then the replay enters on an infinite loop, and the replay got stuck.

This has happened to me with traces of the Unreal Engine 4 Shooter Demo, recorded on the Raspberry4. I tested  the replay with two Mesa vulkan drivers, the one for the Raspberry4 and the Intel on my laptop, and it happens the same in both.

The straightforward hack is to remove the loop, and just call the function once. Doing that we get warnings like this:

> [gfxrecon] WARNING - API call vkGetEventStatus returned value VK_EVENT_RESET that does not match return value from capture file: VK_EVENT_SET.

but the trace continues, so that is better that getting stuck.

Not sure what would be the correct solution, perhaps a limit of calls?  a timeout?

Another interesting issue here is to know why we never get back the status value that we expect, but I assume that is related to the capture.

**Comment by dustin-lunarg** [comment]
For this case, I think we can limit the number calls to resolve the issue.  I have created PR #523 to limit the number of retries.

I have also run into the infinite loop issue when capturing DirectX games on Linux through a DX to Vulkan translation layer, but I have never really investigated the issue thoroughly.  I have always just applied the same hack that you described to remove the loop, intending to implement a proper solution at a later time.

The replay has always worked with the loop removed, producing the same warnings described in the original issue, but I believe we had also received a report with vktrace that not waiting for the event status to match was causing a problem.  So, I think that keeping the loop with a limited number of retries is a good compromise.

One scenario that I think could lead to the infinite loop would be with a multi-threaded application, where one thread waits on vkGetEventStatus to return a specific value before a second thread calls vkQueueSubmit with a command buffer that will set the event status.  With our single threaded replay, we would end up replaying the calls in the order they were captured, where we replay and wait on vkGetEventStatus to return a value that will never be set because replay will not advance to replay the vkQueueSubmit call.

Another scenario that could lead to the problem would be related to the fact that replay doesn't know which status it is waiting for.  The original app may have been waiting for VK_EVENT_RESET, where its first call to vkGetEventStatus returned VK_EVENT_SET and a second call returned VK_EVENT_RESET.  On replay, if the first call returns the expected VK_EVENT_RESET, we'd end up waiting forever for the VK_EVENT_SET status that was captured, but which won't be set.

**Comment by infapi00** [comment]
> For this case, I think we can limit the number calls to resolve the issue. I have created PR #523 to limit the number of retries.

That seems like a nice compromise. Thanks!

**Comment by dustin-lunarg** [comment]
Closed with #523
